### PR TITLE
Inline before merging cgus

### PR DESCRIPTION
**Pull request by nnethercote**
Because CGU merging relies on CGU sizes, but the CGU sizes before inlining aren't accurate.

This change doesn't have much effect on compile perf, but it makes follow-on changes that involve more sophisticated reasoning about CGU sizes much easier.

r? @wesleywiser

### CODE REVIEW

**Comment by nnethercote** [comment]
I bet I can fix those terrible results for `issue-46449` by tweaking the minimum CGU size threshold a tiny bit. I'll look into it on Monday.

**Comment by nnethercote** [comment]
I changed the minimum size to 2500, which gave good results locally.

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
Let's try 1800.

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
I added a couple of additional refactoring commits.

File: compiler/rustc_middle/src/mir/mono.rs [hunk 1]
```diff
@@ -320,19 +320,18 @@ impl<'tcx> CodegenUnit<'tcx> {
...
-        self.size_estimate = Some(self.items.keys().map(|mi| mi.size_estimate(tcx)).sum());
+        self.size_estimate = self.items.keys().map(|mi| mi.size_estimate(tcx)).sum();
     }
 
     #[inline]
-    /// Should only be called if [`create_size_estimate`] has previously been called.
+    /// Should only be called if [`compute_size_estimate`] has previously been called.
     ///
-    /// [`create_size_estimate`]: Self::create_size_estimate
+    /// [`compute_size_estimate`]: Self::compute_size_estimate
```
Comment (wesleywiser) [inline compiler/rustc_middle/src/mir/mono.rs:332]: Nit: you can still detect the case where `compute_size_estimate` was not called before this function was by asserting that if `items.len() > 0`, then `size != 0`. The minimum possible size of a `MonoItem` is `1` so it is impossible for `size_estimate` to actually be `0` if there are any items.

**Review by wesleywiser — APPROVED** [review]
lgtm!

File: compiler/rustc_middle/src/mir/mono.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_middle/src/mir/mono.rs:332]: Good suggestion. I will add `assert!(self.items.is_empty() || self.size_estimate != 0);`.

**Comment by nnethercote** [comment]
I addressed the nit.

@bors r=wesleywiser

**Comment by nnethercote** [comment]
Not sure if this is an intermittent crash, or if my changes have triggered something latent. Lets try again

@bors retry

**Comment by nnethercote** [comment]
The instructions results here look bad, but this was expected. Walltimes are less bad, though a little worse than the pre-merge CI run, which is annoying. This change is a necessary prerequisite for follow-up work to improve CGU formation: CGU merging is entirely driven by CGU size estimates, but when inlining happens after merging, those estimates are unavoidably inaccurate.

@rustbot label: +perf-regression-triaged

**Pull request merged**